### ISSUE TITLE
[FEA] [SID] Add decorder-only sequential gr example 

### ISSUE THREAD

**Issue by JacoCheung** [NVIDIA/recsys-examples#240]
This is a sub-task of #196 .
We will add a SID-based GR example. In this example, we will directly use mcore to implement all parallelism ([**THD** format](https://github.com/NVIDIA/TransformerEngine/blob/main/tests/pytorch/attention/run_attention_with_cp.py#L91) ) and torchrec for embedding.

**What we want to deliver / highlight:**
1. Support jagged input which is typical in RecSys and the corresponding **context parallelism**.  #7 , #207. 
2. Add cuda graph usage example when input is jagged for inference.
3. [Optional] Beam search and cuda graph for inference. 
---
**What we don't include:** 
1. Text encoder -- We assume the input data is preprocessed sid, not multi-modal or text data.
2. Semantic ID quantizations -- We will directly use 3rd lib such as GRID to generate sid. 
---

The (planned) step-by-step actions for T5 are: 
- [x] Refactor existing hstu example, extract shareable codes for SID series model. 
- [x] Define a dataset format which consists of semantic ids and user history and corresponding dataloader. 
- [x] Implement gpt/llama model modules with mcore ~~that aligns with GRID~~.
- [ ] Add training example with #207 #7 .
- [ ] Add inference and beam search.

**Comment by JacoCheung** [comment]
Fixed some bugs regarding attention mask generation logics.

Now the model could be overfit given single batch. However, for normal training, the metric is still not expected. Metric degrades when the sid hierachies for evalution increase.

<img width="3564" height="2064" alt="Image" src="https://github.com/user-attachments/assets/8a6d9aab-7e54-42c8-9e08-2485b3bf628f" />
